### ISSUE TITLE
npm "list" function doesn't check packages.json

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29199]
From @mlococo on 2015-04-08T19:35:00Z

##### Issue Type
Bug Report

##### Component Name
npm module

##### Ansible Version
2.0

##### Steps to reproduce
1. Create /managed/npm/path/package.json containing any valid package.json content.
2. Run an ansible task containing npm: path=/managed/npm/path state=latest
3. Update /managed/npm/path/package.json to specify a new package.
4. Rerun ansible.

##### Expected outcome

The new package is installed

##### Actual outcome

It's not

##### Comments
1. On the first run, npm list --json returns without any "dependencies" section. In the "list" function, "None" is appended to the missing packages list and the install function is run.
2. On the second run, npm list --json returns the "dependencies" which are installed, but omits the newly specified item in package.json. No named packages are passed to the module. Ansible erroneously concludes that there is no work to do.
3. The npm-module's "list" function should explicitly check the content of the package.json found inside the "path". If it contains any items that are omitted from npm-list's output, npm-install should be run.
4. Running npm install outside of ansible results in the package being installed.

Copied from original issue: ansible/ansible-modules-extras#383

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

[module: packaging/language/npm.py]

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@mlococo Thanks for submitting this bug report. Apologies for taking so long to triage it. Since this bug has been open for a while, we would appreciate if you can verify if you are still seeing the reported behavior in the latest version.

If you are seeing the reported behavior, just let us know and we will leave the bug open and notify the module maintainer.

If you are no longer seeing the reported behavior, please let us know and we will close the bug report.

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

I haven't re-verified, but npm.py hasn't had any code-changes (only formatting and attribution commits) since this was reported so it's a pretty fair bet that it is still valid: https://github.com/ansible/ansible-modules-extras/commits/devel/packaging/language/npm.py

Also, the list function is still relying entirely on the npm list output, which I believe (as noted in the original report) is insufficient to detect newly added dependecies. I believe they can only be detected by inspecting the contents of package.json directly and comparing that to the npm list output. https://github.com/ansible/ansible-modules-extras/blob/7f245e8c73e36e8276d87714d4c86a66bfd42019/packaging/language/npm.py#L163

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

Initially I missed the option "state=latest" and it behaved the same way. After adding state=latest, it worked and installed new added libraries in package.json

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@AmirMeirbekov That looks like an artifact of this change: https://github.com/ansible/ansible-modules-extras/commit/7d8dd6e21051cb548c731fc85f76deb980dc66a8#diff-6b9e5546ff117496c9905e1bab64cd63

Looks to me like the bug will still be present for states other than latest, though pragmatically that's probably a reasonable workaround for a large number of cases.

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@mlococo I think this is basically a dup of #957 so you may want to just update your npm with 'npm install npm -g' and/or try the code on PR #1663

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansibot Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- issue type
- ansible version
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @mlococo on 2015-04-08T19:35:00Z

@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @chrishoffman
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @xcambar
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by xcambar** [comment]
Hi,
Can the issue be reproduced with a recent ansible (2.4 or devel)?

As the new maintainer of the npm module, I am flooded by year-old issues. So as a triage strategy, any silent issue will be closed after 10 days unless you confirm your issue still exists in recent versions.

**Comment by mlococo** [comment]
@xcambar I don't have a test case for this handy anymore, but I did some scrolling through the history of npm.py and I doubt this is fixed in the general case. 

1. I believe this change fixed this issue in cases where the state=latest: https://github.com/ansible/ansible-modules-extras/commit/7d8dd6e21051cb548c731fc85f76deb980dc66a8#diff-6b9e5546ff117496c9905e1bab64cd63
2. For other states, I suspect the issue is still present and my expectation is that a change to the list function is required to address the issue, which hasn't happened since 2015 when I last commented.

**Comment by xcambar** [comment]
Thanks for giving a look at this.
I'm leaving this open in hope you can try to reproduce the scenario.

> I don't have a test case for this handy anymore

How did you work around it?

**Comment by mlococo** [comment]
@xcambar As noted in previous comments, state=latest has become a viable workaround in cases where that behavior is desired since I originally filed.

Honestly, it's been so long (almost 2.5 years) I don't remember exactly how I moved passed this problem. I'm automating the setup of a development environment which has build-scripts as well. I suspect that the build scripts also manage npm dependencies and I had ansible install npm deps primarily to aggregate as much slow initial download into one coffee-break as possible... so it was not critical if ansible missed a thing or two as the build-script would grab the missing dependencies on-demand. It's been a while, but that's probably it... this bug was not on the critical path for my use-case but I wanted to file it since I had characterized it pretty specifically by the time I figured out it wasn't critical to me.
